Add pybind support for all memory config options in OrtArenaCfg

petermcaughan opened the pull request:
**Description**: Improve the `pybind` implementation of `OrtArenaCfg` object. Currently, the `OrtArenaCfg` constructor provided through `pybind` doesn't expose all options to the user, and the public class members of `OrtArenaCfg` aren't accessible through `pybind`. This PR makes the following changes:
 - Add new `pybind` constructor for `OrtArenaCfg` that takes a collection of Key-Value pairs instead of ordered parameters. This allows users to access all memory arena options going forward without having to modify the constructor each time a new parameter is added.
 - Add `pybind` support for accessing the class members within `OrtArenaCfg`. These members are public in the C api, and should be accessible through Python as well.
 - UT for changes


**Motivation and Context** 

- There is one config value `initial_growth_chunk_size_bytes` that isn't exposed to the user through the `pybind` offering of `OrtArenaCfg`. See diff between:

   - [`OrtArenaCfg` Constructor](https://github.com/microsoft/onnxruntime/blob/ac7538b909990f89f9bfebb325cd53c3d31cf158/include/onnxruntime/core/framework/allocator.h#L33)
  - [`pybind` implementation](https://github.com/microsoft/onnxruntime/blob/aa85092b51ced97507d7fec12c4c67d6d376aac7/onnxruntime/python/onnxruntime_pybind_state.cc#L1091)

   I don't have reasoning as to why we hide the parameter from users, so this PR aims to add more flexibility to ORT going forward by adding a more flexible constructor and member access.

pranavsharma reviewed (commented):
Changes look good. Also, update https://github.com/microsoft/onnxruntime/blob/df5ee6aa4ee24172fe21dc07c4a76ebc9ed6f54c/onnxruntime/test/python/onnxruntime_test_python.py#L1157 since it shows only 4 arguments.

hariharans29 commented:
> Changes look good. Also, update
> 
> https://github.com/microsoft/onnxruntime/blob/df5ee6aa4ee24172fe21dc07c4a76ebc9ed6f54c/onnxruntime/test/python/onnxruntime_test_python.py#L1157
> 
> since it shows only 4 arguments.

I was wondering if it may be better to take the C API approach of taking in key value pairs so that we don't need to keep updating this ctor each time we add a new config knob (https://github.com/microsoft/onnxruntime/blob/aa85092b51ced97507d7fec12c4c67d6d376aac7/include/onnxruntime/core/session/onnxruntime_c_api.h#L2640) - Maybe like a static factory method for this same class ? 
 
Also, doesn't this change break backwards compat ?

petermcaughan commented:
Updated relevant comments and added overloaded constructor that utilizes Python Dict for initialization - should maintain backwards compatibility and be more flexible going forward. 

Also changed the scope of this PR a bit to also add pybind access to some public class members of OrtArenaCfg. It seems like these should be accessible through Python and makes UT for new constructor possible.
